daemon/cmd/daemon.go

```diff
@@ -555,7 +555,7 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 			k := key.(*ipcachemap.Key)
 			v := value.(*ipcachemap.RemoteEndpointInfo)
 			nid := identity.NumericIdentity(v.SecurityIdentity)
-			if nid.HasLocalScope() {
+			if nid.HasLocalCIDRScope() {
 				d.restoredCIDRs = append(d.restoredCIDRs, k.Prefix())
 				oldNIDs = append(oldNIDs, nid)
```

Comment:
I think that if we push the remote identity restore logic into this restore logic, we can drop the additional costly ipcache restore iteration, remove all of the new nodemanager logic, and remove the extra new identity allocation+release path (+ associated changes to the core agent wait times)

```suggestion
			if nid.HasLocalCIDRScope() || (nid.Scope() == identity.IdentityScopeRemoteNode && option.Config.PolicyCIDRSelectsNodes)
			{
				d.restoredCIDRs = append(d.restoredCIDRs, k.Prefix())
				oldNIDs = append(oldNIDs, nid)
```

I'll point out corresponding changes elsewhere in my next comment which could facilitate this in a relatively minimal way while only slightly tweaking the ipcache internal logic which already has to deal with identity details around specific remote node labels.

(Though we do need to think carefully about restore cases here for restoring various permutations of versions, such as from main today without the feature to this PR, from this PR to this PR restarting, or when toggling the new flag on/off)

Reply:
^ I believe that this small change should automatically enqueue the release of the identities per the existing logic for identity restore & release.

This won't directly associate the remote-node labels with the identity, but we can infer that by masking the identity in the ipcache below. Either way, soon enough the k8s watcher in node manager will also associate that identity.

Reply:
Sure, I can look in to this. I'm not as confident in all the details of the existing code path -- from my experience the older code is a lot more fragile, since it doesn't make the reference counting explicit -- but it should be possible.

Reply:
> we do need to think carefully about restore cases here for restoring various permutations of versions

So, I've given this some thought. The interesting case is transitioning from "enabled" to "disabled", with a policy that selects `reserved:remote-node`. (Transitioning from "disabled" to "enabled" doesn't have any restoration implications; just that some IPs change identity -- something we already handle gracefully).


Restoring to an older version *or* transitioning to the disabled flag currently is almost identical:
1: bpf ipcache map has nodeip `(N -> A)`; bpf policymap selects `(A, 6, 7)` because 6 and 7 are *always* selected)
2: The k8s listeners are synced. The ipcache updates the bpf map to point `(N -> 6)`
3: Endpoints are regenerated: bpf code is updated. In the downgrade case, we now only understand id 6 and 7 to be remote nodes
4: BPF policymap is updated. It now only selects `(6, 7)`

Unfortunately, there is no barrier between `ipcache.InjectLabels()` succeeding and endpoint regeneration, so we are at some risk for dropped traffic between 2 and 3. This *probably* affects other subsystems as well, FWIW.

There might be some other synchronization that ensures the ipcache has finished injecting before we regenerate endpoints, but I couldn't find it.

Reply:
Update: I'm silly. We make a new ipcache map on restart every time. This should certainly be safe to upgrade / downgrade / change flags at will.